src/core/thread/network_data_publisher.cpp

```diff
@@ -0,0 +1,1079 @@
+    static_assert(4 == kRemoving, "kRemoving value is not correct");
+
+    return kStateStrings[aState];
+}
+
+#if OPENTHREAD_CONFIG_TMF_NETDATA_SERVICE_ENABLE
+
+//---------------------------------------------------------------------------------------------------------------------
+// Publisher::DnsSrpServiceEntry
+
+Publisher::DnsSrpServiceEntry::DnsSrpServiceEntry(Instance &aInstance)
+    : mCallback(nullptr)
+    , mCallbackContext(nullptr)
+{
+    Init(aInstance);
+}
+
+void Publisher::DnsSrpServiceEntry::SetCallback(DnsSrpServiceCallback aCallback, void *aContext)
+{
+    mCallback        = aCallback;
+    mCallbackContext = aContext;
+}
+
+void Publisher::DnsSrpServiceEntry::PublishAnycast(uint8_t aSequenceNumber)
+{
+    otLogInfoNetData("Publisher: Publishing DNS/SRP service anycast (seq-num:%d)", aSequenceNumber);
+    Publish(Info::InfoAnycast(aSequenceNumber));
+}
+
+void Publisher::DnsSrpServiceEntry::PublishUnicast(const Ip6::Address &aAddress, uint16_t aPort)
+{
+    otLogInfoNetData("Publisher: Publishing DNS/SRP service unicast (%s, port:%d)", aAddress.ToString().AsCString(),
+                     aPort);
+    Publish(Info::InfoUnicast(kTypeUnicast, aAddress, aPort));
+}
+
+void Publisher::DnsSrpServiceEntry::PublishUnicast(uint16_t aPort)
+{
+    otLogInfoNetData("Publisher: Publishing DNS/SRP service unicast (ml-eid, port:%d)", aPort);
+    Publish(Info::InfoUnicast(kTypeUnicastMeshLocalEid, Get<Mle::Mle>().GetMeshLocal64(), aPort));
+}
+
+void Publisher::DnsSrpServiceEntry::Publish(const Info &aInfo)
+{
+    bool registerWithLeader = false;
+
+    if (GetState() != kNoEntry)
+    {
+        if (aInfo == mInfo)
+        {
+            otLogInfoNetData("Publisher: %s is already being published", ToString().AsCString());
+            ExitNow();
+        }
+
+        registerWithLeader = Remove(/* aNextState */ kNoEntry);
+    }
+
+    mInfo = aInfo;
+    SetState(kToAdd);
+
+    Process();
+
+    if (registerWithLeader)
+    {
+        Get<Notifier>().HandleServerDataUpdated();
+    }
+
+exit:
+    return;
+}
+
+void Publisher::DnsSrpServiceEntry::Unpublish(void)
+{
+    bool registerWithLeader;
+
+    otLogInfoNetData("Publisher: Unpublishing DNS/SRP service");
+
+    registerWithLeader = Remove(/* aNextState */ kNoEntry);
+
+    if (registerWithLeader)
+    {
+        Get<Notifier>().HandleServerDataUpdated();
+    }
+}
+
+bool Publisher::DnsSrpServiceEntry::HandleNotifierEvents(Events aEvents)
+{
+    bool registerWithLeader = false;
+
+    if ((GetType() == kTypeUnicastMeshLocalEid) && aEvents.Contains(kEventThreadMeshLocalAddrChanged))
+    {
+        mInfo.SetAddress(Get<Mle::Mle>().GetMeshLocal64());
+
+        if (GetState() == kAdded)
+        {
+            // If the entry is already added, we need to update it
+            // so we remove it and add it back immediately with
+            // the new mesh-local address.
+
+            Remove(/* aNextState */ kAdding);
```

Comment:
Combine SRV_DATA.ntf with below Add() which happens soon as agreed on https://github.com/openthread/openthread/pull/6587#discussion_r646270190
```suggestion
            Remove(/* aNextState */ kAdding, /* aRegisterWithLeader*/ false);
```

Reply:
> Combine SRV_DATA.ntf with below Add() which happens soon as agreed on #6587 (comment)

@librasungirl  I think this is already the case with code  (i.e. there will be a single `SRV_DATA.ntf` after both `Remove()` and `Add()` calls - which is done from `Publisher::HandleNotifierEvents()`).

Note that `Remove()` and `Add()` no longer have `aRegisterWithLeader` parameter and always expect the caller to take care of it (they have a `bool` return value  `registerWithLeader` to indicate if something was actually changed in the local netdata). This is different and changed from the code in https://github.com/openthread/openthread/pull/6587#discussion_r646270190